lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -962,6 +997,88 @@ static LogicalResult buildAllocOp(ComponentLoweringState &componentState,
+  bool isFloat = !memtype.getElementType().isInteger();
+
+  auto shape = allocOp.getType().getShape();
+  int totalSize =
+      std::reduce(shape.begin(), shape.end(), 1, std::multiplies<int>());
+  // The `totalSize <= 1` check is a hack to:
+  // https://github.com/llvm/circt/pull/2661, where a multi-dim memory whose
+  // size in some dimension equals 1, e.g. memref<1x1x1x1xi32>, will be
+  // collapsed to `memref<1xi32>`, whose `totalSize` equals 1. We should instead
+  // never pass multi-dimensional memories to Calyx, as discussed in:
+  // https://github.com/calyxir/calyx/issues/907
+  if (!(shape.size() <= 1 || totalSize <= 1)) {
+    allocOp.emitError("input memory dimension must be empty or one.");
+    return failure();
+  }
+
+  std::vector<uint64_t> flattenedVals(totalSize, 0);
+  if (isa<memref::GetGlobalOp>(allocOp)) {
+    auto getGlobalOp = cast<memref::GetGlobalOp>(allocOp);
+    auto *symbolTableOp =
+        getGlobalOp->template getParentWithTrait<mlir::OpTrait::SymbolTable>();
+    auto globalOp = dyn_cast_or_null<memref::GlobalOp>(
+        SymbolTable::lookupSymbolIn(symbolTableOp, getGlobalOp.getNameAttr()));
+    // Flatten the values in the attribute
+    auto cstAttr = llvm::dyn_cast_or_null<DenseElementsAttr>(
+        globalOp.getConstantInitValue());
+    int sizeCount = 0;
+    for (auto attr : cstAttr.template getValues<Attribute>()) {
+      assert((isa<mlir::FloatAttr, mlir::IntegerAttr>(attr)) &&
+             "memory attributes must be float or int");
+      if (auto fltAttr = dyn_cast<mlir::FloatAttr>(attr)) {
+        flattenedVals[sizeCount++] =
+            bit_cast<uint64_t>(fltAttr.getValueAsDouble());
+      } else {
+        auto intAttr = dyn_cast<mlir::IntegerAttr>(attr);
+        int64_t value = intAttr.getInt();
+        flattenedVals[sizeCount++] = bit_cast<uint64_t>(value);
+      }
+    }
+
+    rewriter.eraseOp(globalOp);
+  }
+
+  llvm::json::Array result;
+  if (shape.empty()) {
+    llvm::json::Value value = flattenedVals[0];
+    result.emplace_back(std::move(value));
+  } else {
+    for (int i = 0; i < shape[0]; ++i) {
+      result.emplace_back(nullptr);
+    }
+  }
+
+  bool isSignlessOrUnsigned = memtype.getElementType().isSignlessInteger() ||
+                              memtype.getElementType().isUnsignedInteger();
+  for (size_t i = 0; i < flattenedVals.size(); ++i) {
+    uint64_t bitValue = flattenedVals[i];
+    llvm::json::Value value = 0;
+    if (isFloat) {
+      // we cast to `double` and let downstream calyx to deal with the actual
+      // value's precision handling
+      value = bit_cast<double>(bitValue);
+    } else {
+      APInt apInt(/*numBits=*/elmTyBitWidth, bitValue);
+      value =
+          isSignlessOrUnsigned ? apInt.getZExtValue() : apInt.getSExtValue();
+    }
+    result[i] = std::move(value);
+  }
+
+  componentState.setDataField(memoryOp.getName(), result);
+
+  auto width = memtype.getElementType().getIntOrFloatBitWidth();
```

Comment:
Above you call:

```
unsigned elmTyBitWidth = memtype.getElementTypeBitWidth();
``` 

I think that's the same right?

Reply:
sounds good!